ondemand/client/client.go

```diff
@@ -0,0 +1,450 @@
+	resourceName := o.resourceName()
+	if !strings.HasPrefix(m.Type, resourceName) {
+		return
+	}
+
+	var rawQuery ondemand.RawQuery
+	if err := json.Unmarshal(m.Obj, &rawQuery); err != nil {
+		logging.GetLogger().Errorf("unable to decode %s message: %s", resourceName, err)
+		return
+	}
+
+	resource, err := o.handler.DecodeMessage(rawQuery.Resource)
+	if err != nil {
+		logging.GetLogger().Errorf("unable to decode %s: %s", resourceName, err)
+		return
+	}
+
+	switch m.Type {
+	case resourceName + "StartReply":
+		// not registered thus remove from registered cache
+		if m.Status != http.StatusOK {
+			logging.GetLogger().Debugf("%s start request failed %v", resourceName, m.Debug())
+		} else {
+			logging.GetLogger().Debugf("%s start request succeeded %v", resourceName, m.Debug())
+		}
+		// removeRegisteredNode()
+		o.subscriberPool.BroadcastMessage(o.newStructMessage(ondemand.NotificationNamespace, "NodeUpdated", resource.ID()))
+	case resourceName + "StopReply":
+		if m.Status == http.StatusOK {
+			o.removeRegisteredNode(rawQuery.NodeID, resource.ID())
+		} else {
+			logging.GetLogger().Debugf("%s stop request failed %v", resourceName, m.Debug())
+		}
+	}
+}
+
+func (o *OnDemandProbeClient) registerProbes(nodes []interface{}, resource types.Resource) {
+	toRegister := func(node *graph.Node) (nodeID graph.Identifier, host string, register bool) {
+		o.graph.RLock()
+		defer o.graph.RUnlock()
+
+		// check not already registered
+		o.RLock()
+		probes, ok := o.registeredNodes[string(node.ID)]
+		if ok {
+			_, ok = probes[resource.ID()]
+		}
+		o.RUnlock()
+
+		if ok {
+			logging.GetLogger().Debugf("%s already registered on %s", resource.ID(), node.ID)
+			return
+		}
+
+		if !o.handler.CheckNode(node) {
+			return
+		}
+
+		return node.ID, node.Host, true
+	}
+
+	nps := map[graph.Identifier]nodeProbe{}
+	for _, i := range nodes {
+		switch i.(type) {
+		case *graph.Node:
+			node := i.(*graph.Node)
+			if nodeID, host, ok := toRegister(node); ok {
+				nps[nodeID] = nodeProbe{string(nodeID), host, resource}
+			}
+		case []*graph.Node:
+			// case of shortestpath that returns a list of nodes
+			for _, node := range i.([]*graph.Node) {
+				if nodeID, host, ok := toRegister(node); ok {
+					nps[nodeID] = nodeProbe{string(nodeID), host, resource}
+				}
+			}
+		}
+	}
+
+	if len(nps) > 0 {
+		go func() {
+			for _, np := range nps {
+				o.registerProbe(np)
+			}
+		}()
+	}
+}
+
+func (o *OnDemandProbeClient) newStructMessage(ns string, kind string, body interface{}) *ws.StructMessage {
+	return ws.NewStructMessage(ns, fmt.Sprintf("%s%s", o.handler.ResourceName(), kind), body)
+}
+
+func (o *OnDemandProbeClient) registerProbe(np nodeProbe) bool {
+	body, err := o.handler.EncodeMessage(np.id, np.resource)
+	if err != nil {
+		logging.GetLogger().Errorf("Unable to encode message for agent %s: %s", np.host, err)
+		return false
+	}
+
+	msg := o.newStructMessage(ondemand.Namespace, "Start", ondemand.RawQuery{
```

Comment:
I would rather use the namespace to distinguish the endpoint type instead of the resource type + "Start"